src/slurm.c

```diff
@@ -0,0 +1,589 @@
+      return FUTURE_POWERUP;
+    if (power_down_flag)
+      return FUTURE_POWERDOWN;
+    if (no_resp_flag)
+      return FUTURE_NONRESP;
+    return FUTURE;
+  }
+  if (resume_flag)
+    return RESUME;
+  if (base == NODE_STATE_UNKNOWN) {
+    if (no_resp_flag)
+      return UNKNOWN_NONRESP;
+    return UNKNOWN;
+  }
+  return UNKNOWN2;
+}
+
+#define NUM_NODE_STATES (sizeof(node_state_names) / sizeof(node_state_names[0]))
+
+typedef struct partition_state_st {
+  char name[PART_NAME_SIZE];
+  uint32_t nodes_states_count[NUM_NODE_STATES];
+  /* counts jobs states indexed by enum job_states in slurm.h */
+  uint32_t jobs_states_count[JOB_END];
+} partition_state_t;
+
+/* based on enum job_states from slurm.h */
+static const char *job_state_names[] = {
+    "pending", "running",   "suspended", "complete",  "cancelled", "failed",
+    "timeout", "node_fail", "preempted", "boot_fail", "deadline",  "oom",
+};
+
+static partition_state_t *alloc_partition_states(uint32_t num_partitions,
+                                                 partition_info_t *partitions) {
+  partition_state_t *partition_states;
+
+  partition_states =
+      (partition_state_t *)calloc(num_partitions, sizeof(partition_state_t));
+  if (!partition_states) {
+    return NULL;
+  }
+
+  for (int i = 0; i < num_partitions; i++)
+    sstrncpy(partition_states[i].name, partitions[i].name, PART_NAME_SIZE);
+
+  return partition_states;
+}
+
+static partition_state_t *find_partition(partition_state_t *partitions,
+                                         uint32_t num_partitions, char *name) {
+  partition_state_t *part = NULL;
+
+  for (int i = 0; i < num_partitions; i++) {
+    if (strncmp(name, partitions[i].name, PART_NAME_SIZE) == 0)
+      part = &partitions[i];
+  }
+
+  return part;
+}
+
+/*
+ * Submit one gauge value
+ */
+static void slurm_submit(const char *plugin_instance, const char *type,
+                         const char *type_instance, gauge_t value) {
+  value_list_t vl = VALUE_LIST_INIT;
+
+  vl.values = &(value_t){.gauge = value};
+  vl.values_len = 1;
+  sstrncpy(vl.plugin, PLUGIN_NAME, sizeof(vl.plugin));
+  if (plugin_instance != NULL)
+    sstrncpy(vl.plugin_instance, plugin_instance, sizeof(vl.plugin_instance));
+  sstrncpy(vl.type, type, sizeof(vl.type));
+  if (type_instance != NULL)
+    sstrncpy(vl.type_instance, type_instance, sizeof(vl.type_instance));
+
+  plugin_dispatch_values(&vl);
+}
+
+static void slurm_submit_partition(partition_state_t *partition) {
+  for (int i = 0; i < JOB_END; i++) {
+    slurm_submit(partition->name, "slurm_job_state", job_state_names[i],
+                 partition->jobs_states_count[i]);
+  }
+  for (int i = 0; i < NUM_NODE_STATES; i++) {
+    slurm_submit(partition->name, "slurm_node_state", node_state_names[i],
+                 partition->nodes_states_count[i]);
+  }
+}
+
+static void slurm_submit_stats(stats_info_response_msg_t *stats_resp) {
+  slurm_submit("stats", "slurm_stats", "parts_packed",
```

Comment:
The main point that catches me is `slurm_stats` type.

Collectd types has idea what type describes some area, but here we see all data sent in one `slurm_stats` type.
I think, that is wrong and needs more work (and more types).

Good example is `jobs_*` metric, which you put into type instance.
From my POV on Collectd types, that should be own type, possible `slurm_jobs` if no better type found.
And submitted/started/completed/canceled/failed should be type instances.

Install `rrdtool` plugin + setup `collection3` and you will see what are benefits ;-) 
With `slurm_jobs` implemented described way, you can get one chart which shows you all these metrics grouped, and, as I understand, their sum will give us 100% of jobs, so you can see distribution.

Another note - I think you should not set plugin instance value at all. Constant value in that field has no sence, I think.

Reply:
Hi, thanks a lot for the feedback! 

I think you are definitely correct about some of the metrics needing a different RRD type. These are all global types that are not bound to a specific partition, and as you pointed out some make sense to add as DERIVE, also some as COUNTER like *_sum metrics which are monotonically increasing.

However the way in which new types are added is not clear to me. I intended to group metrics together in a way that makes sense, yet does not pollute types.db with lots of new types.

I am not using collection3, but currently the data get funneled through various dbs/frameworks and are displayed with Grafana at the end of the "pipeline", where they are already grouped by their metric name prefix (that is, by their type instance prefix). 

While it would be certainly possible to create a type slurm_job as DERIVE, for the other metrics it is a bit less clear: Not all bf_* metrics and schedule_* metrics are of the same type, and therefore it would not be possible to create a slurm_schedule or slurm_bf type to group them together. Similarly, many other types would need to be created if the goal is to separate/plot them based on the type, in some cases like parts_packed, server_thread_count, agent_queue_size, these would each need a new type (one type per single metric). Unless, they are all grouped together somehow.

So I guess my question is: what granularity should we use to group them vs creating new types? 

IMHO it would make sense to use slurm_stats as plugin instance to group all these global statistics together, and then use the generic types (counter for COUNTER type metrics, derive for DERIVE type metrics, gauge for GAUGE type metrics), and keep the metric names (type instance) as they are.

Does this make sense from a collectd/rrd point of view?

Reply:
I think you should add new types when you think it is needed.
If in doubt, lets discuss.

>counter for COUNTER type metrics, derive for DERIVE type metrics, gauge for GAUGE type metrics

I think that is not good solution, because one should do grouping, discussed above, anyway.

Reply:
> leaving finer granularity grouping up to the plotting software, 

As you already use your plugin, you already do some grouping.
I also think your added grouping will not avoid possibility to do another grouping in plotting software. But presence of that grouping could be useful. So, provide it, please )

>and 3) not polluting types.db with lots of new types.
I dont think that is real problem. My POV is to freely add new types if required.